### PR TITLE
Use std::vector instead of std::map in class ABMHandler, original by Rogier-5

### PR DESCRIPTION
**Pull request by lhofhansl**
Adopting #4895. Done with three commits, the first with @Rogier-5 's original change, a second with my changes on top, to make it clear what I changed. The 3rd commit improves this further, with some suggestions from @gregorycu.

* added the overflow check, avoiding a core dump
* remove the unnecessary in resize
* removed the custom copy constructor for MapNode (now uses the trivial default constructor)
* removed unnecessary logic from MapBlock::getNodeNoEx (check it out it really does exactly the same)
* avoid initializing the vector for the MapNode until needed
* retrieve a MapNode reference directly, a avoid the validity check (thanks @gregorycu)

This help a lot on my system. Would be nice if someone else could verify.

### CODE REVIEW

**Comment by gregorycu** [comment]
Probably the only other suggestion worth making is to actually iterate through each node in the block, as opposed to iterate through the dimensions. Something akin to:

`for (auto node_iter= block->nodeBegin(); node_iter!= block->nodeEnd(); ++node_iter)
{
    content_t c = node_iter->getContent();`

Will lead to less maths in the code, and probably a little bit more of a chance of good cache-locality.

File: src/mapblock.h [hunk 1]
```diff
@@ -513,6 +512,16 @@ class MapBlock /*: public NodeContainer*/
 	void serializeNetworkSpecific(std::ostream &os, u16 net_proto_version);
 	void deSerializeNetworkSpecific(std::istream &is);
 
+	// Use only when the MapBlock is help in scope outside.
+	inline MapNode *getNodeUnsafe(s16 x, s16 y, s16 z)
+	{
+		return data == NULL ? NULL : &data[z * zstride + y * ystride + x];
+	}
+	// Use only when the MapBlock is help in scope outside.
+	inline MapNode *getNodeUnsafe(v3s16 &p)
```
Comment (nerzhul) [inline src/mapblock.h]: missing const

File: src/mapblock.h [hunk 2]
```diff
@@ -513,6 +512,16 @@ class MapBlock /*: public NodeContainer*/
 	void serializeNetworkSpecific(std::ostream &os, u16 net_proto_version);
 	void deSerializeNetworkSpecific(std::istream &is);
 
+	// Use only when the MapBlock is help in scope outside.
+	inline MapNode *getNodeUnsafe(s16 x, s16 y, s16 z)
```
Comment (nerzhul) [inline src/mapblock.h]: i exactly did the same think on my fork to reduce the getNodeNoEx CPU usage

**Comment by gregorycu** [comment]
And with regards to the potentially NULL data pointer, maybe it's worth checking to see if the block is a dummy block first (where data is NULL), and remove the check from (effectively) inside the loop. I have no idea what a dummy block is, but somewhat feels like something that should be checked up-front, as opposed to every use. I know it seems like a bit of micro optimising, but I think it makes the code clearer in any case.

File: src/mapblock.h (hunk 2 above)
Comment (gregorycu) [inline src/mapblock.h]: Great minds etc.

**Comment by lhofhansl** [comment]
How about this?
Leaks a bit of the internals - so I suspect some discussion, but is the fastest I can come up with, and the ABM processing is a pretty core part of what MT is doing, so leaking some abstractions here should be OK.

**Comment by lhofhansl** [comment]
Wait... I got ahead of myself in latest. That can't work right. Will undo sorry.

**Comment by lhofhansl** [comment]
The iterator or direct enumeration does not work (without extra math) can't work since we still need the position of the node in the block.

**Comment by lhofhansl** [comment]
OK... This should be good for review.

File: src/mapblock.h (hunk 1 above)
Comment (lhofhansl) [inline src/mapblock.h]: Fixed in latest.

**Comment by gregorycu** [comment]
I think that's as fast as you're going to make it. What's the perf stats like?

I'll take a look later tonight at the performance, and the generated ASM for windows.

Any further optimisation of ABM would probably involve at looking at what functions the ABMs call and making sure they are optimised. That's likely to be quite a task on its own.

I missed the complexity around the iterator. Sorry about that. Having said that, looping through the ABMs is the common case, and executing the ABM is somewhat rare (the place where you need the coordinates). But probably not worth the hassle.

**Comment by nerzhul** [comment]
percentage depend on the ABM loaded per server now and their complexity

**Comment by gregorycu** [comment]
Yeah, I think our baseline situation (from the previous PR) is minimal subgame, doing nothing.

I have my own stats, I'll take this and compare this against base and the first PR.

**Comment by Zeno-** [comment]
Personally I think there should be a commit in there somewhere with Rogier-5 as the author. But that's just an opinion. Looking good

**Comment by lhofhansl** [comment]
Agree. We can the first commit on behalf of Rogier-5. I prefer to give credit we credit is due.

**Comment by nerzhul** [comment]
Then can you amend first commit, please ?

File: src/mapblock.h [hunk 3]
```diff
@@ -513,6 +512,16 @@ class MapBlock /*: public NodeContainer*/
 	void serializeNetworkSpecific(std::ostream &os, u16 net_proto_version);
 	void deSerializeNetworkSpecific(std::istream &is);
 
+	// Use only when the MapBlock is help in scope outside.
```
Comment (sfan5) [inline src/mapblock.h]: i'm having a hard time understand this, can you clarify?

Comment (asl97) [inline src/mapblock.h]: Might I suggest moving the two function to line 356 where the other getNode function are.
Specifically, just below setNodeNoCheck.
Maybe with `// non-validating variants of the above` instead for the comment.
____

@stan5: I believe it's something along the line of `It should only be use when the MapBlock has been loaded by another function in the same scope/function`.

**Comment by nerzhul** [comment]
We can do it yes but if you do it, it permits to flag pr as merged

**Comment by sfan5** [comment]
:+1:

**Comment by sfan5** [comment]
ad10b8b7 ca362963

**Pull request closed without merging**